Documentation and examples for the new charts interface

rothnic opened the pull request:
The intent of this pr is to address the missing documentation and examples that are broken, or are still using the legacy charts, and can be converted over. This will clean things up a bit and make sure all the base functionality is there before I implement the remaining charts.

Only modifications to charts are those that must exist to fix or implement the examples.

This will close issue #2975.

rothnic commented:
Screenshot of the scatter.py examples with added one with negative values:

![image](https://cloud.githubusercontent.com/assets/452052/10469036/26b6341e-71c8-11e5-81f7-5098c6630ffa.png)

Shows the example including negative values for bar docstring:

![image](https://cloud.githubusercontent.com/assets/452052/10469074/76c1006a-71c8-11e5-816b-2237d5e6f2ec.png)

rothnic commented:
Updated stacked bar chart example in gallery:

![image](https://cloud.githubusercontent.com/assets/452052/10469193/23851f34-71c9-11e5-9445-f38b62a625cc.png)

fpliger commented:
thanks @rothnic ! I was planning to review this PR today but let me know if you think I should wait until you turn it to `ready` status.

Are you starting to convert the legacy charts? Otherwise I can start working on it.

fpliger commented on bokeh/charts/_attributes.py:
```diff
@@ -197,8 +206,41 @@ def color(columns=None, palette=None, **kwargs):
+        **kwargs: any keyword, arg supported by :class:`AttrSpec`
+
+    Returns:
+        :class:`CatAttr`
```
`MarkerAttr`? :)

fpliger commented on bokeh/charts/_builder.py:
```diff
@@ -106,21 +109,26 @@ class Builder(HasProps):
+
+    attributes = Dict(String, Instance(AttrSpec), help="""The attribute specs used to
+        group data.""")
+    default_attributes = Dict(String, Instance(AttrSpec), help="""The attribute specs
```
I think we could improve the help text in order to make it clear about the difference between attributes and default_attributes

rothnic commented:
@fpliger , thanks for the review. There are some implementation details I've been conflicted on that we should probably go over before diving in. I'm planning to work tomorrow some to finish this PR up.

fpliger commented:
@rothnic awesome. I'm still looking on it (have been having lots of interruptions)... are you planning on including some work to port the last legacy charts? If not I'd be happy to start working on those.

rothnic commented:
@fpliger: I'm only doing documentation in this PR, but my next task is to port over the remaining charts. The TimeSeries one should be super simple, since Line pretty much does it already. The others will be more work.

The tests passed on my local machine, I'll wait to see what travis reports and do another scan through before marking it ready.

fpliger commented on bokeh/charts/attributes.py:
```diff
@@ -197,8 +278,41 @@ def color(columns=None, palette=None, **kwargs):
+        columns (list or str): the columns used to generate the categorical variable
+        cats (list, optional): overrides the values derived from columns
+        sort (bool, optional): whether to sort the values
+        **kwargs: any keyword, arg supported by :class:`AttrSpec`
```
ascending docstring missing

fpliger commented on bokeh/charts/builder.py:
```diff
@@ -73,30 +76,26 @@ class Builder(HasProps):
+      comp_glyphs for their minimum and maximum values. See the
+      :meth:`~bokeh.charts.builder.Builder.create` method for more information on
+      when this is called and how the builder provides the ranges to the containing
+      :class:`Chart` using the :meth:`Chart.add_ranges` method.
```
small thing, I'd list the mandatory methods first and the optional ones later (it may sound a bit picky, sorry :)

fpliger commented on bokeh/charts/data_source.py:
```diff
@@ -214,22 +298,49 @@ def get_selections(self, selections, **kwargs):
     def __getitem__(self, dim):
         """Get the columns selected for the given dimension name.
 
+        If the dimension is not in `_selections`, `None` is returned.
```
I'd move this to the Returns section

fpliger commented on bokeh/charts/data_source.py:
```diff
@@ -214,22 +298,49 @@ def get_selections(self, selections, **kwargs):
+
+        A wide table is one where the column names represent a categorical variable
+        and each contains only the values associated with each unique value of the
+        categorical variable.
```
IMO this docstring is a bit misleading.. because it only talks about the table wide->tall conversion and not about melting the data with the current datagroup _data...

Also a simple example would be useful

fpliger commented on bokeh/charts/glyphs.py:
```diff
@@ -1,25 +1,30 @@
-    """A composite glyph that consists of other composite glyphs."""
+    """A composite glyph that consists of other composite glyphs.
+
+    An important responsibility of the `CompositeGlyph` is to understand the bounds
```
`CompositeGlyph` -> `NestedCompositeGlyph`?

fpliger commented on bokeh/charts/utils.py:
```diff
@@ -269,24 +232,20 @@ def df_from_json(data, **kwargs):
-    return pd.Series(data.index.values)
+    """A generic function to return the index from values.
+
+    Should be used to abstract away from specific types of data.
```
We could add something about what are the supported input types (`data`)

fpliger commented:
@rothnic I left some minor comments that could be sticked in or easily be part of a future PR. It's definitely a great improvement. :+1:

rothnic commented:
@fpliger: think I got them taken care of.

One thing remaining... Do I need to replace the gallery image thumbnail manually or is that auto generated by some process I'm not aware of?

fpliger commented:
@rothnic Awesome! 

Regarding the gallery, IIRC they are generated and should be handled by https://github.com/bokeh/bokeh/blob/master/sphinx/main_gallery.json

rothnic commented:
If they should be produced automatically, then something might not be working correctly. When building the docs I see the images copied from source to _build, but the source images are of the old chart.

We can see this in the current release gallery:
Chart: http://bokeh.pydata.org/en/latest/docs/gallery/boxplot_chart.html
Thumbnail:
![image](https://cloud.githubusercontent.com/assets/452052/10611411/934f8f3c-7711-11e5-8641-d2045672ec8c.png)

That specific example probably should be more in line with what is shown in the thumbnail, but either way, it seems to not be updated at the moment.

fpliger commented:
@rothnic sorry, I think my memory is failing here.. You are right. Thumbs should be in https://github.com/bokeh/bokeh/tree/master/sphinx/source/_images/gallery

rothnic commented:
Ok, wanted to make sure. I'll add them.

rothnic commented:
All gallery examples have been updated and thumbnails for them updated, so marking this ready.

Only two things really touch outside of charts, besides the documentation.

* I added intersphinx to sphinx with references to python, numpy, and pandas documentation. This supports autolinking direct to external documentation when using something like ```:class:`pandas.DataFrame````.
* I moved the charts default palette to plotting helpers at the request of a ToDo that was next to it for some time. I could rename this to charts default palette if that would be required.

bryevdv commented:
I don't think the `_gallery` is needed on the example filenames, no other examples do that. The `main_gallery.json` is specifically for mapping arbitrary example filenames, so there's no need to add an extraneous bits to the filename.

rothnic commented:
@bryevdv, the _gallery was more to identify those examples from the ones that go through the many different input combinations. I thought for the gallery it would be good to just have a single good example,  which also works better with the thumbnails.

fpliger commented:
+1 on removing `_gallery`. Maybe use `_simple` or name `_multi` the others showing different use cases in the same example?

It'd be good to avoid the duplication since those `_gallery` are duplication needed just for the gallery but can't think of a better solution atm.

Apart from that LGTM

rothnic commented:
@fpliger: renamed them. 

There really is unneeded duplication between the file examples and the notebooks example to begin with that I don't care for. In the future I'd like to look at leveraging the notebooks more to walkthroughs. I [have a modified sphinx extension](https://github.com/rothnic/sphinx_notebook) that allows you to include notebooks into documentation as is or by cell and executes them during the build process.

fpliger commented:
Cool! we could make good use of that...

fpliger commented on bokeh/compat/mplexporter/exporter.py:
```diff
@@ -189,7 +189,7 @@ def draw_line(self, ax, line, force_trans=None):
                 or markerstyle['markerpath'][0].size == 0):
             markerstyle = None
-        label = line.get_label()
+        label = line._get_label()
```
why?

rothnic replied:
Seems I must have misapplied a refactor on that file. Just fixed it and pushed the change. I checked the remaining files to make sure nothing else was like that
